### PR TITLE
Update to Hadoop 3.3.5

### PR DESCRIPTION
**Pull request by oneonestar**
supersede #37 

## Changes
src/main/java/io/trino/hadoop/SocksSocketFactory.java
- guava from hadoop-common upgraded from 11.0 to 27.0
- getHostText() has been renamed to getHost()
- Ref: https://github.com/google/guava/blob/a35bfd9ea855539354b783ee74839e384f7af98a/guava/src/com/google/common/net/HostAndPort.java#L91

src/main/java/org/apache/hadoop/crypto/key/kms/KMSClientProvider.java
- Apply HADOOP-17288: Use shaded guava from org.apache.hadoop.thirdparty.

src/main/java/org/apache/hadoop/fs/FileSystem.java
- HADOOP-17313: FileSystem.Cache became a final class. A semaphore has being added to getInternal().
- Looks like it is safe to remove the `final` keyword

src/main/java/org/apache/hadoop/fs/ForwardingFileSystemCache.java
- The constructor now requires a Configuration()
- Ref: https://github.com/apache/hadoop/blob/a3b9c37a397ad4188041dd80621bdeefc46885f2/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FileSystem.java#L206

src/main/java/org/apache/hadoop/util/LineReader.java
* Copy from Hadoop 3.3.4
* Apply patches  6622830, 15c26aa, e57f61b

io/trino/hadoop/TestHadoopNative.java
- See HADOOP-17125.

## Removed files
src/main/java/org/apache/hadoop/security/authentication/util/KerberosUtil.java
- The existing version is 3.2.0 + HADOOP-10848 + HADOOP-17432
- Both patches are included in 3.3.1

src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java
- HADOOP-16479 is included in 3.3.1

src/main/java/org/wildfly/openssl/OpenSSLProvider.java
- Changes in response to HADOOP-16371

## pom.xml
Update slf4j to 1.7.36 in order to align with hadoop
https://github.com/apache/hadoop/blob/a585a73c3e02ac62350c136643a5e7f6095a3dbb/hadoop-project/pom.xml#L81

Add jsr305 for javax.annotation for JDK > 9

Add a dependency to org.lz4:lz4-java
> HADOOP-17292: The lz4-java is declared in provided scope. Applications that wish to use lz4 codec must declare dependency on lz4-java explicitly.

The version is 1.7.1 in order to align with hadoop-project.
https://github.com/apache/hadoop/blob/a585a73c3e02ac62350c136643a5e7f6095a3dbb/hadoop-project/pom.xml#L146

Add a provided dependency to org.xerial.snappy:snappy-java.
This is caused by HADOOP-17125.
Since snappy-java is already a dependency for Trino, scope=provided is enough for the test to pass.
The version is 1.1.8.2 in order to align with hadoop-project.
https://github.com/apache/hadoop/blob/a585a73c3e02ac62350c136643a5e7f6095a3dbb/hadoop-project/pom.xml#L145

Add relocations for the new dependencies come from hadoop

## TODO
- [ ] Update the binaries inside src/main/resources/nativelib
- [ ] Test with different environments

### CODE REVIEW

**Comment by oneonestar** [comment]
@zielmicha I'll be appreciated if you could help on testing this PR.
I just bumped the version from 3.3.1 to 3.3.4. I'm going to test the patch with Hadoop 2.7 and 3.3.
AFAIK Hadoop doesn't provide any explicit promise on compatibility across major versions.
Any additional tests would help a lot.


Also, it looks like all the binaries need to built by a project maintainer (https://github.com/trinodb/trino-hadoop-apache/pull/27#issuecomment-780307327). We'll need some help from @electrum 
I would also like to know what's the current status of CI tests coverage on different versions of Hadoop.

**Comment by zielmicha** [comment]
I made some small fixes to make main Trino repo compile here:
https://github.com/zielmicha/trino-hadoop-apache/commits/hadoop_3.3.4

I tried to test it using test Trino/Hive cluster and I'm getting weird errors - NoClassDefFoundError for `UserGroupInformation$HadoopConfiguration`, even though this file exists in jar and other classes (e.g. `UserGroupInformation`) load fine. Any ideas what might be going wrong?

**Comment by oneonestar** [comment]
## javax.annotation
https://github.com/zielmicha/trino-hadoop-apache/commit/4ff5ab6aea4f4dd312c78d291f04d9653ca8456f#diff-9c5fb3d1b7e3b0f54bc5c4182965c4fe1f9023d449017cece3005d3f90e8e4d8R597-R600
`javax.annotation` only contains annotations. I'm not sure is it necessary to relocate it.


## okhttp
https://mvnrepository.com/artifact/com.squareup.okhttp/okhttp
`com.squareup.okhttp` has been renamed to `com.squareup.okhttp3`
I make a update in bf4a04d

## kotlin-stdlib
```
[INFO] +- org.apache.hadoop:hadoop-mapreduce-client-core:jar:3.3.4:runtime
[INFO] |  \- org.apache.hadoop:hadoop-hdfs-client:jar:3.3.4:runtime
[INFO] |     +- com.squareup.okhttp3:okhttp:jar:4.9.3:runtime
[INFO] |     |  \- com.squareup.okio:okio:jar:2.8.0:runtime
[INFO] |     +- org.jetbrains.kotlin:kotlin-stdlib:jar:1.4.10:runtime
[INFO] |     \- org.jetbrains.kotlin:kotlin-stdlib-common:jar:1.4.10:runtime
```
Somehow Kotlin has been included in `hadoop-hdfs-client`. It looks like an unintentional change to me: https://github.com/apache/hadoop/pull/4229#issuecomment-1179963242
Other projects are also being affected: https://github.com/apache/hbase/pull/4687/files
Hopefully they will fix the dependency soon: https://issues.apache.org/jira/browse/HDFS-16714
I shaded the Kotlin related lib in bf4a04d

## KerberosUtil
KerberosUtil has been removed since the patches are already included in 3.3.1
I forgot to remove the exclusion in pom.xml which caused a runtime error.
This has been fixed in 0883143.

---

@zielmicha I have successfully make queries on Hadoop 3.3.x cluster.
 I can't reproduce your error. Could you try the new changes and see if it works for you?

**Comment by zielmicha** [comment]
javax.annotation:
Did you build the main Trino repo with the new dependency or just replaced the jars? I think not shading works fine at runtime, but Maven complains about duplicate classes.

**Comment by oneonestar** [comment]
Ok. I added the relocation rule for javax.annotation.

I removed `LdapGroupsMapping.java` because #32 should already be fixed by
https://github.com/apache/hadoop/commit/f257497b0f964c171961be4ad527983f24f9be31

**Comment by zielmicha** [comment]
After the last round of changes, I no longer see any errors at runtime. My guess is this problem was related to KerberosUtil (`UserGroupInformation$HadoopConfiguration` uses it), the error message was just super unclear.

I'll do few more manual tests (note the configuration I have does not use HDFS, so I don't really test "against" Hadoop, just confirming that the libraries work okay internally). I'll also run the regression tests suite. 

The okhttp3 rule I added is also needed to build the main repo.

**Comment by zielmicha** [comment]
I've made another small change to pom.xml: https://github.com/zielmicha/trino-hadoop-apache/commit/ca49d402788baa9a5d4d7aba7377f978ab1af9a8

The regression tests in the main repo now pass.

**Comment by oneonestar** [comment]
```
$ jar tf target/hadoop-apache-3.3.4-1-SNAPSHOT.jar | grep okhttp3
okhttp3/
okhttp3/Request$Builder.class
okhttp3/Dispatcher.class
okhttp3/Headers$Companion.class
okhttp3/MultipartBody$Part.class
okhttp3/Protocol.class
...
```

You are right. `<pattern>okhttp3</pattern>` is the correct way.

**Comment by zielmicha** [comment]
I tested the current version on a small test cluster with Hive connector (using NFS for storage, not HDFS). Not sure what other testing we should do, maybe it makes sense to request review from maintainers now?

**Comment by oneonestar** [comment]
Sure. Let's go for it.

**Comment by zielmicha** [comment]
cc @electrum

**Comment by KarlManong** [comment]
@oneonestar @zielmicha io.trino.rcfile.TestRcFileReader failed "java.lang.NoClassDefFoundError: org/xerial/snappy/Snappy"

**Comment by oneonestar** [comment]
@KarlManong Thanks for the info.
I created a PR in Trino to handle the issue: https://github.com/trinodb/trino/pull/14674
I'm going to verify is `snappy-java` needed for `trino-orc`.

File: pom.xml
Comment (ebyhr) [inline pom.xml:651]: I think the subsequent filter for `org.apache.hadoop:hadoop-azure` needs update. I got a class not found when running `TestDeltaLakeAdlsConnectorSmokeTest` locally.

**Comment by mladjan-gadzic** [comment]
@oneonestar hello! Is there any motion on this PR? There are erros [HadoopIllegalArgumentException: Invalid buffer, not of length X" when querying Hive erasure coding tables](https://github.com/trinodb/trino/issues/6413 ) when using Trino with Apache Hadoop version 3.2.0. Error mentioned before is fixed in Apache Hadoop version 3.3.0 [EC : Decoding is failing when block group last incomplete cell fall in to AlignedStripe](https://issues.apache.org/jira/browse/HDFS-14373) however, Trino is still using old version of Apache Hadoop. Is there any way we can push this forward?

File: pom.xml
Comment (oneonestar) [reply pom.xml:651]: I have updated the filter for `org.apache.hadoop:hadoop-azure`.
However, I can't run `TestDeltaLakeAdlsConnectorSmokeTest` because I don't have access to azure-abfs.

https://github.com/trinodb/trino/blob/master/plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeAdlsConnectorSmokeTest.java#L64-L66

**Comment by oneonestar** [comment]
@mladjan-gadzic We faced the same problem before. We applied this patch to our internal Trino cluster and it works fine with EC.

There are a few problems we have to solve before this PR can be merged:
* Help from a committer who can compile the necessary [nativelibs](https://github.com/trinodb/trino-hadoop-apache/tree/master/src/main/resources/nativelib)
* Find a way to run all Trino integration tests with the updated `trino-hadoop-apache` to ensure this won't break things accidentally
* HDFS-16453 upgraded okhttp from 2.7.5 to 4.9.3. This brings the whole kotlin-stdlib into transitive dependency. We have to accept this or find a way to work around it. (Currently I shaded the kotlin lib)

**Comment by mladjan-gadzic** [comment]
@oneonestar thank you for a quick answer!

> * Help from a committer who can compile the necessary [nativelibs](https://github.com/trinodb/trino-hadoop-apache/tree/master/src/main/resources/nativelib)

Unfortunately I am using macos with m1 chip which introduces whole new level of issues when tinkering around architecture dependent stuff. Because of this I am unable to compile native libs. But what I can do is I can check if someone from my organization can do that.

> * Find a way to run all Trino integration tests with the updated trino-hadoop-apache to ensure this won't break things accidentally

I will try to do this and get back to you.

> * HDFS-16453 upgraded okhttp from 2.7.5 to 4.9.3. This brings the whole kotlin-stdlib into transitive dependency. We have to accept this or find a way to work around it. (Currently I shaded the kotlin lib)

What are the pros and cons of accepting kotlin-stdlib as transitive dependency? Are there any downsides for shading aside what shading brings to the table itself?

**Comment by mladjan-gadzic** [comment]
Hi @oneonestar! Just a remainder to check up my [comment](https://github.com/trinodb/trino-hadoop-apache/pull/44#issuecomment-1422873080). I am eager to help push this PR forward.

**Comment by dave-gantenbein** [comment]
@oneonestar @electrum - any chance we could get some eyes on this? Thanks in advance!

**Comment by KarlManong** [comment]
hadoop 3.3.5 has been released at 2023-03-15: https://hadoop.apache.org/docs/stable/hadoop-project-dist/hadoop-common/release/3.3.5/CHANGELOG.3.3.5.html

**Comment by c-rindi** [comment]
Hello @electrum, @oneonestar, @ebyhr - checking in - any progress?

**Review by electrum — APPROVED** [review]

**Comment by electrum** [comment]
@oneonestar apologies for the very long delay on this. I updated your branch to use Hadoop 3.3.5 and copied the Hadoop native libraries from the official Hadoop releases (they have `aarch64` now). I didn't bother to rebuild the other versions since there are only minimal changes which shouldn't affect testing on macOS. The tests passed in https://github.com/trinodb/trino/pull/17869

**Pull request merged**
